### PR TITLE
Pod count as load metrics in BurstMetrics Service

### PR DESCRIPTION
**Pull request by kforeverisback**
# Type of PR

- [X] Code changes

## PR Checklist

- [X] All new and existing tests passed.
- [X] My code follows the code style of this project.
- [X] I ran lint checks locally prior to submission.
- [X] Threw exception manually to test `204-NoContent`

## Purpose of PR

Change the load type from CPU metrics to Pod count from HPA.

## Does this introduce a breaking change

- [X] YES
- [ ] NO

Previously BurstMetrics Service returned CPU percentage as load metrics.
With this PR, BurstMetrics will return POD count as load metrics.

## Validation

- [X] Ran local tests and integration tests
- [X] Update documentation or issue referenced above

## Issues Closed

- Closes https://github.com/retaildevcrews/wcnp/issues/280
- Closes https://github.com/retaildevcrews/wcnp/issues/294
- Closes https://github.com/retaildevcrews/wcnp/issues/299

### CODE REVIEW

File: burst/src/Controllers/BurstMetricsController.cs
Comment (jofultz) [inline burst/src/Controllers/BurstMetricsController.cs:45]: In what way are the -1 values intended to be used?  I see no conditional for this value so currently it looks like if no values come back then this method would return a header with all -1 values.  I don't think that is what we want.  I'm okay with the nullables, because it means we don't have to build some complex error handling, but maybe instead of using a string have it return the value and then test for the -1 in any of the values and if any of them are -1 then return a 204 with an empty body.

File: burst/src/K8sApi/K8sHPAMetricsService.cs
Comment (jofultz) [inline burst/src/K8sApi/K8sHPAMetricsService.cs:77]: If an error is caught for any of those value assignments then we need to set all of the values to the invalid value (right now you're using -1).  That way in the final condition before returning we can test and see that we should return a 204.  Also, create a constant for invalid value, maybe something like, `private const int INVALID_METRIC_VALUE` then you can use that for assignment and test and we have a single place to change it if we need to change it.

File: burst/src/Controllers/BurstMetricsController.cs
Comment (kforeverisback) [reply burst/src/Controllers/BurstMetricsController.cs:45]: I agree, as part of https://github.com/retaildevcrews/wcnp/issues/294, check if any value is not available (i.e. null), then return 204 without any content.

We could check if the metrics are null (since they are nullable)

```C#
# pseudo code
if (hpaMetrics?.TargetLoad == null ||
    hpaMetrics?.CurrentLoad == null || 
    hpaMetrics?.MaxLoad == null)
{
    return <HTTP 204 status with empty body>;
}
else
{
    <previous code block with 200-content goes here>
}
```

Thoughts?

File: burst/src/K8sApi/K8sHPAMetricsService.cs
Comment (kforeverisback) [reply burst/src/K8sApi/K8sHPAMetricsService.cs:77]: `hpaMetrics` members will remain null without any value assignmentsm since it's initialized at line 51.

Do we still need to set MaxLoad, CurrentLoad and TargetLoad to an `INVALID_METRIC_VALUE` to check in BurstMetricsService?

File: burst/src/Controllers/BurstMetricsController.cs
Comment (jofultz) [reply burst/src/Controllers/BurstMetricsController.cs:45]: Whether it's null or -1 is a decision that needs to be made then what we do with it is the next part. We can take care of what to do with it in the next PR, but I'm this PR you're assigning the value so we should decide which value to use and why.

File: burst/src/K8sApi/K8sHPAMetricsService.cs
Comment (jofultz) [reply burst/src/K8sApi/K8sHPAMetricsService.cs:77]: Depends on how you test. If the first and second assignment succeed but the third fails then you have a two valid values that will be returned. If, especially under maintenance, the check is not correct then you could end up sending back partially correct data.

File: burst/src/K8sApi/K8sHPAMetricsService.cs
Comment (kforeverisback) [inline burst/src/K8sApi/K8sHPAMetricsService.cs:93]: @jofultz What happens if we can't find the HPA? Say, the deployment or the namespace doesn't exist. Or we don't have RBAC permission in that specific NS/Deployment.

Do we still return `null` from this function?
In that case controller will return 204 no-content.

Should we consider 404 for no-hpa found?

**Comment by bartr** [comment]
- blocked until after Spark Encore presentations

**Review by jofultz — APPROVED** [review]

**Pull request merged**
